### PR TITLE
Fixed to_future completion on empty Observables

### LINKED ISSUE #325: Infinite Busy Wait
# Reproducing

```python
import asyncio
import rx

async def x(y):
    return await y
    
y = rx.Observable.from_([])
asyncio.get_event_loop().run_until_complete(y)
```

# Tested Versions

Tested on RxPy 1.6, but also confirmed the issue is present on master.

# Potential Fixes

[The following implementation](https://github.com/ReactiveX/RxPY/blob/b333440871ad5e74dbdf5365fe99643f22f92c25/rx/core/operators/tofuture.py#L33) shows the issue: if no value has been produced, awaiting the observable will block forever, since no result or exception is set.

In my (biased) opinion, a few options are possible, since I feel that silently blocking forever based on an opaque type is an unsatisfactory solution:

1. Silently set the value to `None`. This would have the added advantage that the original statement would work, but would also be difficult to distinguish between an function that does not return a value, and an empty iterable, which are two completely different types. It also has 1 more benefit, which is similar to future-like objects (see below).
2. Raise `StopAsyncIteration`. This would be explicit, but would mean the above statement would fail, but would clearly signal intent.
3. Raise a custom error. Also explicit, and would also clearly signal intent. It could potentially derive from `StopAsyncIteration`.

# Footnotes

**Future-Like Objects**

[PEP 492](https://www.python.org/dev/peps/pep-0492/#asyncio) defines future-like objects which have `__await__ = __iter__` to simplify implementations. These cannot return a value, and I believe are only used for the side-effects (that is, to ensure the logic in `__iter__` is executed without having to modify existing code).

```python
import asyncio

class A:
    def __iter__(self):
        return self
    def __next__(self):
        raise StopIteration
    __await__ = __iter__

asyncio.get_event_loop().run_until_complete(A())
```

### PR DESCRIPTION
**Pull request by MainRo**
- Set the future result to None if no item is emitted before completion
- Add more information in documentation
- Replace array with nonlocal

This PR fixes #325.

### CODE REVIEW

**Comment by dbrattli** [comment]
Hi, I think the right thing to do is to cancel the future. This should schedule the callbacks and raise `CancelledError`. Returning `None` for an empty Observable would give the same result as a singleton observable with the value None, e.g `rx.of(None)`.

There is however a comment in the code that no cancellation can be done, but I don't remember why and it might anyways be old. If cancelling doesn't work then we should set an exception.

From the C# impl:

```c#
            public override void OnCompleted()
            {
                if (_hasValue)
                {
                    _tcs.TrySetResult(_lastValue);
                }
                else
                {
                    _tcs.TrySetException(new InvalidOperationException(Strings_Linq.NO_ELEMENTS));
                }

                _ctr.Dispose(); // no null-check needed (struct)
                Dispose();
            }
```

**Comment by jcafhe** [comment]
Don't know if it helps, but maybe we can call `future.cancel()` if there's no items.
```python
        def on_completed():
            if future_result:
                future.set_result(future_result)
            else:
                future.cancel()
```

With the example in #325, the console stops with a CancelledError:
```python
Traceback (most recent call last):
  File "C:\DEV\pyProjects\rx_issues\issue_325\raw.py", line 8, in <module>
    asyncio.get_event_loop().run_until_complete(y)
  File "C:\anaconda3\envs\rxdev\lib\asyncio\base_events.py", line 568, in run_until_complete
    return future.result()
concurrent.futures._base.CancelledError
```

**Comment by MainRo** [comment]
I chose to complete the future with a None value because I had one use case in mind: Using an observable as a lifetime for a future. This is the kind of thing I use from time to time: Apply the ignore_elements on a hot observable, and use the resulting observable completion event as the end of a lifetime. In this case I would expect that the future completes when the observable completes (with no value since I don't care about it).
But if we consider the behavior of a future, then something that ends without a value is... well not possible. The C# implementation has the advantage of considering this case as an error, so there is no ambiguity. 
The to_future operator provides the future, so the only actions that can be done on the future is to complete it with either a result or an exception. Cancelling the future should only be done by the code that received the future from to_future. Moreover if the future is cancelled by the caller, then the to_future implementation should handle this and unsubscribe from the observable (by registering a done_callback and checking if the future is cancelled).

So I would go for do the following implementation:
- set_exception if no item has been emitted on completion.
- handle future cancellation by registering a done callback where the source observable is unsubscribed if the callback is called due to cancellation.

With this my initial use-case can be implemented with an additional concat on an just observable.

File: rx/core/operators/tofuture.py [hunk 1]
```diff
@@ -21,17 +25,18 @@ def to_future(source: Observable) -> Future:
...
+            nonlocal future_result
+            future_result = value
 
         def on_error(err):
             future.set_exception(err)
 
         def on_completed():
-            if has_value:
-                future.set_result(has_value.pop())
+            nonlocal future_result
```
[the remark this answers is not in the document]
Comment (MainRo) [reply rx/core/operators/tofuture.py]: no, I will remove it.

**Comment by MainRo** [comment]
I pushed a new proposal. 
I know why there is the no cancellation comment: The future is created from __await__ and it seems it is not possible to cancel an await, so the futures created here cannot be cancelled.

**Comment by jcafhe** [comment]
I was thinking that maybe it would be usefull to use the sentinel value `rx.core.basics.NotSet` instead of `None` to detect an absence of value. This would allow `None`to be considered as a valid value.

> I pushed a new proposal.
I know why there is the no cancellation comment: The future is created from await and it seems it is not possible to cancel an await, so the futures created here cannot be cancelled.

So what about your use case? Is it working anyway for you? 
Maybe we can parameterize the returned Observable to get the expected behaviour (error or default value).

**Comment by dbrattli** [comment]
Agree that we should throw exception and not cancel. But perhaps we should use `SequenceContainsNoElementsError` instead of `ValueError`? Don't think we should use a sentinel value as code like `next(iter([]))` throws `StopIteration` exception instead of returning something.

**Comment by MainRo** [comment]
So we consider an empty observable as a valid case and we complete the future with the result set to rx.internal.basic.NotSet. Is this ok ?
 I'm fine with both proposals (result NotSet or exception SequenceContainsNoElementsError).

**Comment by dbrattli** [comment]
I think the idea is that the result should never be set to `NotSet`. You only use it internally for `future_result` to know if you have received a value or not. So it should be exception `SequenceContainsNoElementsError` anyways.

**Comment by jcafhe** [comment]
Absolutly, something like:
```python
future_result = NotSet
...

if future_result is NotSet:
    ...
```

**Comment by MainRo** [comment]
ok, here is a new proposal. Sorry in advance, it is not exactly what you both proposed: I think I stricktly sticked to the c# implementation by using two variables. The reason is that we cannot forbid to emit NotSet in an observable. So if we use NotSet, there is the same issue than with using None: to_future will fail with rx.of(NotSet) although it is much less probable to happen. 
With the two variables there is no corner case. However if you prefer, I can change this to use NotSet as a sentinel value.

**Pull request merged**